### ISSUE TITLE
After upgrade to 3.8.3 config, an error occurs with old nightly Neovim versions

### ISSUE THREAD

**Issue by mei28** [lukas-reineke/indent-blankline.nvim#936]
### Problem

After upgraded to 3.8.3, this happens

```

Failed to run `config` for indent-blankline.nvim
.../share/nvim/lazy/indent-blankline.nvim/lua/ibl/utils.lua:22: type: expected function: 0x0100a91ac0, got string (HIGHLIGHT_SETUP)
# stacktrace:
  - vim/shared.lua:0 _in_ **validate**
  - /indent-blankline.nvim/lua/ibl/utils.lua:22 _in_ **validate**
  - /indent-blankline.nvim/lua/ibl/hooks.lua:61 _in_ **register**
  - ~/.config/nvim/lua/plugin/indent.lua:25 _in_ **config**

```

### Steps to reproduce

```
  {
    "lukas-reineke/indent-blankline.nvim",
    main = "ibl",
    version = 'v3.8.3',
    event = { 'BufNewFile', 'BufRead' },
    config = function()
      vim.opt.list = true
      vim.opt.listchars:append "space: "
      -- vim.opt.listchars:append "eol:↴"
      local highlight = {
        "RainbowRed",
        "RainbowYellow",
        "RainbowBlue",
        "RainbowOrange",
        "RainbowGreen",
        "RainbowViolet",
        "RainbowCyan",
      }

      local hooks = require "ibl.hooks"
      -- create the highlight groups in the highlight setup hook, so they are reset
      -- every time the colorscheme changes
      hooks.register(hooks.type.HIGHLIGHT_SETUP, function()
        vim.api.nvim_set_hl(0, "RainbowRed", { fg = "#E06C75" })
        vim.api.nvim_set_hl(0, "RainbowYellow", { fg = "#E5C07B" })
        vim.api.nvim_set_hl(0, "RainbowBlue", { fg = "#61AFEF" })
        vim.api.nvim_set_hl(0, "RainbowOrange", { fg = "#D19A66" })
        vim.api.nvim_set_hl(0, "RainbowGreen", { fg = "#98C379" })
        vim.api.nvim_set_hl(0, "RainbowViolet", { fg = "#C678DD" })
        vim.api.nvim_set_hl(0, "RainbowCyan", { fg = "#56B6C2" })
      end)

      local status, ibl = pcall(require, 'ibl')
      if not status then
        return
      end
      ibl.setup {
        indent = {
          char = "¦",
          highlight = highlight,
        },
      }
    end,
  },

```

### Expected behavior

"The settings worked well in v3.8.2, but an error occurs in v3.8.3."

### Neovim version (nvim -v)

NVIM v0.11.0-nightly+852954f

**Comment by calvinchoy** [comment]
Not sure what the correct solution is, but you can temporary comment out:

``` 
indent = {
          char = "¦",
          highlight = highlight,
        },
 ```
 
 to use the default indent options and prevent the error from showing up.

**Comment by exzolink** [comment]
same for `hooks.register(hooks.type.SCOPE_HIGHLIGHT, hooks.builtin.scope_highlight_from_extmark)`

**Comment by nifoc** [comment]
I guess the faster version of `vim.validate` is causing this

cc @MariaSolOs

**Comment by sinkr** [comment]
Same here.

**Comment by mei28** [comment]
did it cause by this commit #934  ?

**Comment by dzintars** [comment]
Same. For now just tagged previous version.

```lua
return {
  'lukas-reineke/indent-blankline.nvim',
  tag = 'v3.8.2',
  enabled = true,
  main = 'ibl',
  opts = {},
  dependencies = {
    { 'HiPhish/rainbow-delimiters.nvim', lazy = true },
  },

  config = function(_)
    local highlight = {
      'RainbowRed',
      'RainbowYellow',
      'RainbowBlue',
      'RainbowOrange',
      'RainbowGreen',
      'RainbowViolet',
      'RainbowCyan',
      'CursorColumn',
      'WhiteSpace',
    }
    local hooks = require 'ibl.hooks'
    hooks.register(hooks.type.HIGHLIGHT_SETUP, function()
      vim.api.nvim_set_hl(0, 'RainbowRed', { fg = '#E06C75' })
      vim.api.nvim_set_hl(0, 'RainbowYellow', { fg = '#E5C07B' })
      vim.api.nvim_set_hl(0, 'RainbowBlue', { fg = '#61AFEF' })
      vim.api.nvim_set_hl(0, 'RainbowOrange', { fg = '#D19A66' })
      vim.api.nvim_set_hl(0, 'RainbowGreen', { fg = '#98C379' })
      vim.api.nvim_set_hl(0, 'RainbowViolet', { fg = '#C678DD' })
      vim.api.nvim_set_hl(0, 'RainbowCyan', { fg = '#56B6C2' })
    end)

    vim.g.rainbow_delimiters = { highlight = highlight }

    require('ibl').setup {
      indent = {
        highlight = highlight,
        char = '┊',
        tab_char = '|',
      },
      scope = {
        enabled = true,
        show_start = false,
        show_end = false,
        highlight = highlight,
      },
      whitespace = {
        remove_blankline_trail = false,
      },
      exclude = {
        filetypes = {
          'NvimTree',
          'Trouble',
          'dashboard',
          'git',
          'help',
          'markdown',
          'notify',
          'packer',
          'sagahover',
          'terminal',
          'undotree',
        },
        buftypes = { 'terminal', 'nofile', 'prompt', 'quickfix' },
      },
    }
  end,
}
```

**Comment by BigJayToDaIzo** [comment]
> Same. For now just tagged previous version.
> 
> ```lua
> return {
>   'lukas-reineke/indent-blankline.nvim',
>   tag = 'v3.8.2',
>   enabled = true,
>  ...
> }
> ```
Not to cloud the thread with 'me too!''s but adding the tag like the OP is a tremendous temp fix till it's sorted! :rocket:

**Comment by guilhas07** [comment]
Can't reproduce this problem with the current nightly version. When using nightly, make sure to update it frequently.

**Comment by MariaSolOs** [comment]
Sorry for breaking you guys, but as @guilhas07, this shouldn't happen if you're using the tippy top of nightly!

**Comment by nifoc** [comment]
> Sorry for breaking you guys, but as @guilhas07, this shouldn't happen if you're using the tippy top of nightly!

Thanks for clarifying.

I understand it's not your problem, but because of other changes nix/NixOS users can't update to the latest nightly atm.

**Comment by sinkr** [comment]
My LazyBim installation is just the default behavior; no version, tag, etc.

Whatever is going on shouldn't break default behavior.

**Comment by MariaSolOs** [comment]
> My LazyVim installation is just the default behavior; no version, tag, etc.
> 
> Whatever is going on shouldn't break default behavior.

I mean your Neovim version, not `ibl` version. Lazyvim doesn't install Neovim for you.

**Comment by MariaSolOs** [comment]
> I understand it's not your problem, but because of other changes nix/NixOS users can't update to the latest nightly atm.

@nifoc I didn't mean for my comment to sound as if "this isn't my problem", because it is. I didn't want to break anyone's setup but fix a deprecation in one of my favorite plugins.

This should only happen if you have nvim 0.11 (something after the latest release) installed but you're not keeping up to date with the latest nightly. If you're using any nvim version after the latest stable, it _should_ be what's on master.

I do understand that this isn't always possible though, so I'll try to fix it later today after my day job.

**Comment by dzintars** [comment]
I personally don't mind to tag previous version until I will decide to rebuild my nightly. I'm just lazy today to do the build. :) So... no issue for me at least.

**Comment by nifoc** [comment]
> I didn't mean for my comment to sound as if "this isn't my problem", because it is.

@MariaSolOs It absolutely didn't, sorry if my comment came across as "passive aggressive" - it absolutely wasn't intended to be.

I just wanted to add a comment for other Nix users that might come across this issue. The nightly overlay is currently stuck on a roughly two week old commit, because some changes in the way `lpeg` is built (on macOS only?) have to be adapted/worked around.

In no way do I expect you to find a workaround for this. "Normal" nightly users can just update to the latest commit and be done with it.

**Comment by MariaSolOs** [comment]
@nifoc thank you for the clarification. I also apologize if I was a bit aggressive with my reply. I truly apologize for disrupting people's workflows and I'll do my best to address the people having issues.

Still glad to know that there's a workaround.

**Comment by lukas-reineke** [comment]
We maybe should have waited a bit with this until places like Nix are up to date, sorry about that. I don't use Nix, I wasn't aware of this issue.

But generally, **only the latest nightly build is supported!**

If you are using nightly, I would always recommend to only update plugins and Neovim together.
If you don't want to have to deal with issues like this, use stable Neovim instead.

**Comment by MariaSolOs** [comment]
@lukas-reineke I again apologize for causing all this trouble 😅

**Comment by lukas-reineke** [comment]
It's not your fault, all good.
I pinned the issue. We don't need to change the code.

**Comment by mei28** [comment]
Thank you for the discussion. I was able to resolve this issue. I'll wait for the nightly version update in Nix. 

@lukas-reineke  Should I close this issue, or would it be better to keep it open?

**Comment by lukas-reineke** [comment]
Let's keep it open until Nix is caught up with the nightly build.

**Comment by BigJayToDaIzo** [comment]
![image](https://github.com/user-attachments/assets/6aab86d0-e284-41df-b214-cf0976462e95)
![image](https://github.com/user-attachments/assets/cf2105e3-64c0-45d8-bd55-cca96710e24c)

I was under the impression that the `neovim-git` AUR package was building off of nightly, but I pulled the tag out of the plugin table that was holding it back to it's 3.8.2 state and it's still tossing errors at me.  So I guess Arch users are sitting right alongside Nix waiting on our package managers to get caught up.

**Comment by lukas-reineke** [comment]
AUR `-git` packages don't work like normal packages. The version is not updated every time upstream changes, but when you build the package, it will download the latest version from GitHub.  
You just need to explicitly install it again.

**Comment by BigJayToDaIzo** [comment]
> AUR `-git` packages don't work like normal packages. ...
> https://wiki.archlinux.org/title/Arch_User_Repository#What_is_the_difference_between_foo_and_foo-git_packages%3F

I've been using Arch and the AUR for over a year and was today years old when this nuance finally clicked, with your very astute assistance. :grin: Thank you for bridging the gap in my understanding!  Great repo by the way! <3

**Comment by muhammadzkralla** [comment]
Is this fixed yet? If not, is there a temporary fix for it?

**Comment by mei28** [comment]
@muhammadzkralla 
A temporary fix way is to downgrade.
```
return {
  'lukas-reineke/indent-blankline.nvim',
  tag = 'v3.8.2',
```

**Comment by lukas-reineke** [comment]
Looks like Nix has fixed their issue https://github.com/NixOS/nixpkgs/pull/351423

Everyone should be able to update now, so I'll close this.

TLDR for everyone seeing this later

## Only Neovim stable, and the _latest_ Neovim nightly build are supported

**Comment by Kilill** [comment]
This is still and issue on **Ubuntu 24.04** since it chips with `Neovim NVIM v0.11.0-dev` 
Quick fix is either, as mentioned degrade to previous version, or the quick and dirty way:
change the the "1" in the version test (` if vim.fn.has "nvim-0.11" == 1 then`) on line 17  in `indent-blankline.nvim/lua/ibl/utils.lua` to... anything else like "2"

**Comment by lukas-reineke** [comment]
Neovim ships its own lua runtime, your systems lua version does not matter.  
I am very confident your Neovim version was out of date.

**Comment by kakitkelvinho** [comment]
I thought so as well however I have updated/reinstalled neovim many times to no success and the version is already 0.11 but after updating Lua it seems to be working so 🤷‍♂️
